services:
  marzban:
    image: gozargah/marzban:v0.8.4
    restart: always
    env_file: .env
    network_mode: host
    container_name: marzban-1
# марзбан стартует и без этого, но если не делать healthcheck у бд, то при запуске марзбан может сыпать в логи
# ошибки о том что не удалось подключиться к бд (потом все равно подключается)
    depends_on:
      mariadb:
        condition: service_healthy
    volumes:
      - /var/lib/marzban:/var/lib/marzban
      - marzban-sock:/run/:rw
      - xray1-sock:/opt/xray/:rw
      - nginx1-sock:/opt/nginx/:rw
      - /etc/sysctl.conf:/etc/sysctl.conf
      - db-sock:/opt/mariadb/:rw

  mariadb:
    image: mariadb:lts
    network_mode: host
    restart: always
    container_name: mariadb-1
    command:
      - --skip-networking
      - --character_set_server=utf8mb4
      - --collation_server=utf8mb4_unicode_ci
      - --innodb_snapshot_isolation=OFF
      - --host-cache-size=0
      - --innodb-open-files=1024
      - --innodb-buffer-pool-size=268435456
    volumes:
      - ./mysql/:/var/lib/mysql/:rw
      - db-sock:/var/run/mysqld/:rw
	healthcheck:
      test: ["CMD", "test", "-S", "/var/run/mysqld/mysqld.sock"]
      start_period: 3s
      interval: 1s
      timeout: 1s
      retries: 5

# столкнулся с каким-то не особо воспроизводимым багом у нгинкса (1.29 и 1.28):
# если при запуске нгинкс создает сокет, а потом ругается что он уже есть и не помогают
# даже полное удаление контейнера и волюмов, то выставьте нгинксу слушать на tcp портах
# запустите один раз так (без сокетов), потом включите обратно прослушивание на сокете.
# такая проблема наблюдалась только при создании\пересоздании контейнера. в обычной работе не встречал.
  nginx:
    image: nginx:mainline-alpine
    restart: always
    volumes:
      - ./nginx.conf:/etc/nginx/nginx.conf:ro
      - ./certbot/www/:/var/www/certbot/:ro
      - ./certbot/conf/:/etc/letsencrypt/:ro
      - ./log/:/var/log/nginx/:rw
      - nginx1-sock:/opt/run/:rw
      - marzban-sock:/opt/marzban/:rw
      - /etc/sysctl.conf:/etc/sysctl.conf
    network_mode: host
    container_name: nginx-1

  certbot:
    image: certbot/certbot:latest
    volumes:
      - ./certbot/www/:/var/www/certbot/:rw
      - ./certbot/conf/:/etc/letsencrypt/:rw
    network_mode: host

# volumes для сокетов. если мапит из просто папки на хосте, то в случае аварийного выключения сервера
# nginx и xray не могут стартануть потому что старые файлы сокетов не были удалены.
# db-sock и marzban-sock используются только локально, xray-sock и nginx-sock должны подключаться и к хапрокси
# так что создаются там (не принципиально где, можно хоть руками один раз создать и выставить external: true везде)
volumes:
  db-sock:
    driver: local
    driver_opts:
      type: tmpfs
      device: tmpfs
      o: size=4k,nosuid,nodev

  marzban-sock:
    driver: local
    driver_opts:
      type: tmpfs
      device: tmpfs
      o: size=4k,nosuid,nodev


  xray1-sock:
    external: true

  nginx1-sock:
    external: true
